File: sdks/javascript/src/signing.ts
import * as nacl from 'tweetnacl';
import * as util from 'tweetnacl-util';

/**
 * Ed25519 keypair interface
 */
export interface Ed25519Keypair {
  privateKey: Uint8Array;
  publicKey: Uint8Array;
}

/**
 * Ed25519 KeyPair class for agent signing
 * Provides OOP encapsulation similar to Go SDK
 */
export class KeyPair {
  public readonly publicKey: Uint8Array;
  public readonly privateKey: Uint8Array;

  constructor(keypair: Ed25519Keypair) {
    this.publicKey = keypair.publicKey;
    this.privateKey = keypair.privateKey;
  }

  /**
   * Generate a new Ed25519 keypair
   */
  static generate(): KeyPair {
    const keypair = generateEd25519Keypair();
    return new KeyPair(keypair);
  }

  /**
   * Create keypair from base64-encoded private key
   * @param privateKeyBase64 Base64-encoded private key (64 bytes)
   */
  static fromBase64(privateKeyBase64: string): KeyPair {
    const privateKey = decodePrivateKey(privateKeyBase64);
    const publicKey = privateKey.slice(32); // Last 32 bytes are public key
    return new KeyPair({ privateKey, publicKey });
  }

  /**
   * Create keypair from raw private key bytes
   * Supports both 32-byte seed and 64-byte full private key
   * @param privateKey Private key as Uint8Array
   */
  static fromPrivateKey(privateKey: Uint8Array): KeyPair {
    if (privateKey.length === 32) {
      // 32-byte seed - derive full keypair
      const keypair = nacl.sign.keyPair.fromSeed(privateKey);
      return new KeyPair({
        privateKey: keypair.secretKey,
        publicKey: keypair.publicKey,
      });
    } else if (privateKey.length === 64) {
      // 64-byte full private key
      const publicKey = privateKey.slice(32);
      return new KeyPair({ privateKey, publicKey });
    } else {
      throw new Error(
        `Invalid private key length: expected 32 or 64 bytes, got ${privateKey.length}`
      );
    }
  }

  /**
   * Sign a message
   * @param message Message string to sign
   * @returns Base64-encoded signature
   */
  sign(message: string): string {
    const messageBytes = util.decodeUTF8(message);
    const signedMessage = nacl.sign(messageBytes, this.privateKey);
    const signature = signedMessage.slice(0, nacl.sign.signatureLength);
    return util.encodeBase64(signature);
  }

  /**
   * Verify a signature
   * @param message Original message string
   * @param signatureBase64 Base64-encoded signature
   * @returns True if signature is valid
   */
  verify(message: string, signatureBase64: string): boolean {
    try {
      const messageBytes = util.decodeUTF8(message);
      const signature = util.decodeBase64(signatureBase64);
      const signedMessage = new Uint8Array(signature.length + messageBytes.length);
      signedMessage.set(signature);
      signedMessage.set(messageBytes, signature.length);
      const opened = nacl.sign.open(signedMessage, this.publicKey);
      return opened !== null;
    } catch {
      return false;
    }
  }

  /**
   * Sign a JSON payload
   * @param payload Object to sign
   * @returns Base64-encoded signature
   */
  signPayload(payload: Record<string, any>): string {
    return signRequest(this.privateKey, payload);
  }

  /**
   * Get public key as base64 string
   */
  publicKeyBase64(): string {
    return encodePublicKey(this.publicKey);
  }

  /**
   * Get private key as base64 string
   */
  privateKeyBase64(): string {
    return encodePrivateKey(this.privateKey);
  }

  /**
   * Get 32-byte seed as base64 string
   */
  seedBase64(): string {
    const seed = this.privateKey.slice(0, 32);
    return util.encodeBase64(seed);
  }
}

/**
 * Generate Ed25519 keypair for agent signing
 * @returns Keypair with private and public keys
 */
export function generateEd25519Keypair(): Ed25519Keypair {
  const keypair = nacl.sign.keyPair();
  return {
    privateKey: keypair.secretKey,
    publicKey: keypair.publicKey,
  };
}

/**
 * Sign request data using Ed25519 private key
 * Data is converted to canonical JSON (sorted keys) for consistency
 * @param privateKey Ed25519 private key (64 bytes)
 * @param data Data to sign
 * @returns Base64-encoded signature
 */
export function signRequest(privateKey: Uint8Array, data: Record<string, any>): string {
  // Convert data to canonical JSON (sorted keys)
  const jsonString = JSON.stringify(sortObject(data));
  const message = util.decodeUTF8(jsonString);

  // Sign the message
  const signedMessage = nacl.sign(message, privateKey);

  // Extract signature (first 64 bytes)
  const signature = signedMessage.slice(0, nacl.sign.signatureLength);

  // Return base64-encoded signature
  return util.encodeBase64(signature);
}

/**
 * Verify Ed25519 signature
 * Used for testing and validation
 * @param publicKey Ed25519 public key (32 bytes)
 * @param data Original data
 * @param signatureB64 Base64-encoded signature
 * @returns True if signature is valid
 */
export function verifySignature(
  publicKey: Uint8Array,
  data: Record<string, any>,
  signatureB64: string
): boolean {
  try {
    // Convert data to canonical JSON
    const jsonString = JSON.stringify(sortObject(data));
    const message = util.decodeUTF8(jsonString);

    // Decode signature
    const signature = util.decodeBase64(signatureB64);

    // Reconstruct signed message
    const signedMessage = new Uint8Array(signature.length + message.length);
    signedMessage.set(signature);
    signedMessage.set(message, signature.length);

    // Verify signature
    const opened = nacl.sign.open(signedMessage, publicKey);
    return opened !== null;
  } catch (error) {
    return false;
  }
}

/**
 * Encode public key to base64
 * @param publicKey Ed25519 public key
 * @returns Base64-encoded public key
 */
export function encodePublicKey(publicKey: Uint8Array): string {
  return util.encodeBase64(publicKey);
}

/**
 * Decode base64-encoded public key
 * @param publicKeyB64 Base64-encoded public key
 * @returns Ed25519 public key
 */
export function decodePublicKey(publicKeyB64: string): Uint8Array {
  const decoded = util.decodeBase64(publicKeyB64);
  if (decoded.length !== nacl.sign.publicKeyLength) {
    throw new Error(
      `Invalid public key size: expected ${nacl.sign.publicKeyLength}, got ${decoded.length}`
    );
  }
  return decoded;
}

/**
 * Encode private key to base64
 * @param privateKey Ed25519 private key
 * @returns Base64-encoded private key
 */
export function encodePrivateKey(privateKey: Uint8Array): string {
  return util.encodeBase64(privateKey);
}

/**
 * Decode base64-encoded private key
 * @param privateKeyB64 Base64-encoded private key
 * @returns Ed25519 private key
 */
export function decodePrivateKey(privateKeyB64: string): Uint8Array {
  const decoded = util.decodeBase64(privateKeyB64);
  if (decoded.length !== nacl.sign.secretKeyLength) {
    throw new Error(
      `Invalid private key size: expected ${nacl.sign.secretKeyLength}, got ${decoded.length}`
    );
  }
  return decoded;
}

/**
 * Sort object keys recursively for canonical JSON
 * @param obj Object to sort
 * @returns Sorted object
 */
function sortObject(obj: any): any {
  if (obj === null || typeof obj !== 'object' || Array.isArray(obj)) {
    return obj;
  }

  const sorted: Record<string, any> = {};
  const keys = Object.keys(obj).sort();

  for (const key of keys) {
    sorted[key] = sortObject(obj[key]);
  }

  return sorted;
}

File: sdks/go/signing.go
package aimsdk

import (
	"bytes"
	"context"
	"crypto/ed25519"
	"crypto/rand"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"net/http"
	"time"
)

// KeyPair represents an Ed25519 public/private key pair
type KeyPair struct {
	PublicKey  ed25519.PublicKey
	PrivateKey ed25519.PrivateKey
}

// GenerateKeyPair generates a new Ed25519 keypair
func GenerateKeyPair() (*KeyPair, error) {
	publicKey, privateKey, err := ed25519.GenerateKey(rand.Reader)
	if err != nil {
		return nil, fmt.Errorf("failed to generate keypair: %w", err)
	}

	return &KeyPair{
		PublicKey:  publicKey,
		PrivateKey: privateKey,
	}, nil
}

// NewKeyPairFromPrivateKey creates a KeyPair from an existing private key.
// Supports both 32-byte seed format and 64-byte (seed+public) format for compatibility.
func NewKeyPairFromPrivateKey(privateKeyBytes []byte) (*KeyPair, error) {
	var privateKey ed25519.PrivateKey
	var publicKey ed25519.PublicKey

	switch len(privateKeyBytes) {
	case ed25519.SeedSize: // 32 bytes - seed only
		privateKey = ed25519.NewKeyFromSeed(privateKeyBytes)
		publicKey = privateKey.Public().(ed25519.PublicKey)

	case ed25519.PrivateKeySize: // 64 bytes - seed + public key
		privateKey = ed25519.PrivateKey(privateKeyBytes)
		publicKey = privateKey.Public().(ed25519.PublicKey)

	default:
		return nil, fmt.Errorf("invalid private key length: expected 32 or 64 bytes, got %d", len(privateKeyBytes))
	}

	return &KeyPair{
		PublicKey:  publicKey,
		PrivateKey: privateKey,
	}, nil
}

// NewKeyPairFromBase64 creates a KeyPair from base64-encoded private key string.
// Supports both 32-byte and 64-byte private key formats.
func NewKeyPairFromBase64(privateKeyBase64 string) (*KeyPair, error) {
	privateKeyBytes, err := base64.StdEncoding.DecodeString(privateKeyBase64)
	if err != nil {
		return nil, fmt.Errorf("failed to decode base64 private key: %w", err)
	}

	return NewKeyPairFromPrivateKey(privateKeyBytes)
}

// Sign signs a message with the private key and returns base64-encoded signature
func (kp *KeyPair) Sign(message string) (string, error) {
	if kp.PrivateKey == nil {
		return "", fmt.Errorf("private key is nil")
	}

	messageBytes := []byte(message)
	signature := ed25519.Sign(kp.PrivateKey, messageBytes)

	return base64.StdEncoding.EncodeToString(signature), nil
}

// Verify verifies a signature against a message using the public key
func (kp *KeyPair) Verify(message string, signatureBase64 string) (bool, error) {
	if kp.PublicKey == nil {
		return false, fmt.Errorf("public key is nil")
	}

	signature, err := base64.StdEncoding.DecodeString(signatureBase64)
	if err != nil {
		return false, fmt.Errorf("failed to decode signature: %w", err)
	}

	messageBytes := []byte(message)
	return ed25519.Verify(kp.PublicKey, messageBytes, signature), nil
}

// PublicKeyBase64 returns the public key as base64-encoded string
func (kp *KeyPair) PublicKeyBase64() string {
	return base64.StdEncoding.EncodeToString(kp.PublicKey)
}

// PrivateKeyBase64 returns the private key as base64-encoded string
func (kp *KeyPair) PrivateKeyBase64() string {
	return base64.StdEncoding.EncodeToString(kp.PrivateKey)
}

// SeedBase64 returns only the 32-byte seed portion of the private key as base64
func (kp *KeyPair) SeedBase64() string {
	seed := kp.PrivateKey.Seed()
	return base64.StdEncoding.EncodeToString(seed)
}

// SignPayload signs a JSON payload and returns base64-encoded signature
// This is used for signing registration and verification requests
func (kp *KeyPair) SignPayload(payload map[string]interface{}) (string, error) {
	payloadJSON, err := json.Marshal(payload)
	if err != nil {
		return "", fmt.Errorf("failed to marshal payload: %w", err)
	}

	return kp.Sign(string(payloadJSON))
}

// VerificationResult represents the result of an action verification request
type VerificationResult struct {
	Success        bool                   `json:"success"`
	Verified       bool                   `json:"verified"`
	Message        string                 `json:"message"`
	TrustScore     float64                `json:"trust_score"`
	VerificationID string                 `json:"verification_id"`
	Details        map[string]interface{} `json:"details,omitempty"`
}

// SignMessage is a convenience method that signs a message using the client's keypair
func (c *Client) SignMessage(message string) (string, error) {
	if c.keyPair == nil {
		return "", fmt.Errorf("client has no keypair configured")
	}

	return c.keyPair.Sign(message)
}

// VerifyAction sends a verification request to the AIM backend
// This allows the agent to verify an action before executing it
func (c *Client) VerifyAction(ctx context.Context, actionType string, resource string, context map[string]interface{}) (*VerificationResult, error) {
	if c.keyPair == nil {
		return nil, fmt.Errorf("client has no keypair configured")
	}

	// Create verification payload
	payload := map[string]interface{}{
		"action_type": actionType,
		"resource":    resource,
		"context":     context,
		"timestamp":   fmt.Sprintf("%d", nowUnixMillis()),
	}

	// Sign the payload
	payloadJSON, err := json.Marshal(payload)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal payload: %w", err)
	}

	signature, err := c.keyPair.Sign(string(payloadJSON))
	if err != nil {
		return nil, fmt.Errorf("failed to sign payload: %w", err)
	}

	// Prepare request body
	reqBody := map[string]interface{}{
		"action_type": actionType,
		"resource":    resource,
		"context":     context,
		"signature":   signature,
		"public_key":  c.keyPair.PublicKeyBase64(),
	}

	bodyBytes, err := json.Marshal(reqBody)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal request: %w", err)
	}

	// Send verification request
	url := fmt.Sprintf("%s/api/v1/agents/%s/verify", c.config.APIURL, c.config.AgentID)
	req, err := http.NewRequestWithContext(ctx, "POST", url, bytes.NewReader(bodyBytes))
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", c.config.APIKey))

	resp, err := c.reporter.httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to send request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("verification request failed with status: %d", resp.StatusCode)
	}

	var result VerificationResult
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	return &result, nil
}

// nowUnixMillis returns current Unix timestamp in milliseconds
func nowUnixMillis() int64 {
	return time.Now().UnixNano() / int64(time.Millisecond)
}
